src/cargo/ops/registry.rs

```diff
@@ -55,6 +57,39 @@ pub fn publish(manifest_path: &Path,
     Ok(())
 }
 
+fn check_git_cleanliness(current_path: &Path) -> CargoResult<()>{
+    match git2::Repository::open(current_path){
+        Ok(repo) =>{
+            let mut opts = git2::StatusOptions::new();
+            opts.include_untracked(true).recurse_untracked_dirs(true);
+            let status = try!(repo.statuses(Some(&mut opts)));
+            match status.len(){
+                0 => {
+                    Ok(())
+                },
+                num @ _ => {
+                    Err(human(format!("{} uncommited or untacked files \
+                that need to be addressed before publishing", num)))
+                }
+            }
+        },
+        Err(_) => Err(human("unable to open local git"))
+    }
+}
+
+fn check_directory_cleanliness(config: &Config) -> CargoResult<()>{
+    let allow_dirty = try!(config.get_bool("publish.allow_untracked_changes"))
```

Comment:
This may actually be best configured via a command line flag perhaps? I suspect that this flag would be passed in one-off situations rather than blanket configured for the whole system